[WIP - Do not merge] Add support for System.Private.CoreLib to be core library

gkhanna79 opened the pull request:
This change enables System.Private.CoreLib to be the core library for CoreCLR, instead of mscorlib.dll. Updated build scripts and packaging to carry the NI as well.

@jkotas @AlexGhiondea @weshaggard PTAL.

gkhanna79 commented:
@dotnet-bot test Windows_NT arm64 Checked

jkotas commented on build.cmd:
```diff
@@ -408,15 +408,15 @@ echo %__MsgPrefix%Using environment: "%__VSToolsRoot%\VsDevCmd.bat"
 call                                 "%__VSToolsRoot%\VsDevCmd.bat"
 
 if defined __SkipMscorlibBuild (
```
Are you going through and rename these later?

gkhanna79 replied:
Yes, this is already done.

jkotas commented on src/ilasm/asmman.cpp:
```diff
@@ -319,7 +319,7 @@ void    AsmMan::EmitFiles()
 void    AsmMan::StartAssembly(__in __nullterminated char* szName, __in_opt __nullterminated char* szAlias, DWORD dwAttr, BOOL isRef)
 {
-    if(!isRef && (0==strcmp(szName,"mscorlib"))) ((Assembler*)m_pAssembler)->m_fIsMscorlib = TRUE;
+    if(!isRef && (0==strcmp(szName, CoreLibName_A))) ((Assembler*)m_pAssembler)->m_fIsMscorlib = TRUE;
```
This should stay as mscorlib

gkhanna79 replied:
Fixed.

jkotas commented on src/ilasm/asmman.cpp:
```diff
@@ -378,6 +378,7 @@ void AsmMan::EmitDebuggableAttribute(mdToken tkOwner)
         fOldStyle = (m_pAssembly->usVerMajor == 1);
     else
     {
+        // TODO_SPC: ?
```
Same here - keep it as mscorlib

jkotas replied:
(Everywhere in ilasm/ildasm)

gkhanna79 replied:
Thank you!

jkotas commented on src/mscorlib/src/System/Resources/ManifestBasedResourceGroveler.cs:
```diff
@@ -253,7 +253,12 @@ internal static CultureInfo GetNeutralResourcesLanguage(Assembly a, ref Ultimate
                 {
-                    Contract.Assert(false, "mscorlib's NeutralResourcesLanguageAttribute is a malformed culture name! name: \"" + cultureName + "\"  Exception: " + e);
+#if FEATURE_CORECLR
+                    string sCoreLibName = "System.Private.CoreLib";
```
We should have it as C# constant somewhere to avoid these ifdefs, similar to how you have done it for the unmanaged side.

E.g. 
```
internal class CoreLib { 
#if FEATURE_CORECLR
    public const string Name = "System.Private.CoreLib";
    ... other constants as necessary ...
#else
    public const string Name = "mscorlib";
    ... other constants as necessary ...
#endif
}
```

gkhanna79 replied:
Agreed. I already had the code refactored for this.

jkotas commented on src/mscorlib/src/System/Security/SecurityManager.cs:
```diff
@@ -194,6 +194,7 @@ static public PolicyLevel LoadPolicyLevelFromString(string str, PolicyLevelType
             return LoadPolicyLevelFromStringHelper(str, null, type);
         }
 
+        // TODO_SPC: ?
```
This is not compiled into coreclr. I would leave it alone.

jkotas commented on src/mscorlib/src/System/Threading/Tasks/TPLETWProvider.cs:
```diff
@@ -21,6 +21,7 @@ namespace System.Threading.Tasks
 {
     using System.Diagnostics.Tracing;
 
+    // TODO_SPC: ?
```
Yes, this needs a fix.

gkhanna79 replied:
Fixed.

jkotas commented on src/vm/securityattributes.cpp:
```diff
@@ -1816,6 +1816,7 @@ HRESULT SecurityAttributes::GetDeclaredPermissions(IN IMDInternalImport *pIntern
     return hr;
 }
 
+// TODO_SPC: ?
```
This should stay as mscorlib (also, it is not used in coreclr).

AlexGhiondea commented:
@gkhanna79 do we also want to include the reference assembly in the package?

gkhanna79 commented:
I can but @ericstj was not sure if adding it to ref was the right thing. Did you both close on it?

gkhanna79 commented:
@dotnet-bot test Windows_NT x64 Release Priority 1 Build and Test please

gkhanna79 commented:
@dotnet-bot test Windows_NT x86 legacy_backend Checked Build and Test  please

gkhanna79 commented:
@dotnet-bot test Windows_NT arm64 Checked

gkhanna79 commented:
@AlexGhiondea I have rebased this to be in sync with master. I have also updated model.corelib.xml for https://github.com/dotnet/coreclr/pull/4911/files/08dc81499cfb5b7e58bd366642d7920b5dbb8761#r63294835 (FYI - @tijoytom).

gkhanna79 commented:
Also picked up https://github.com/dotnet/coreclr/commit/6c0426b7b5b2529eeb39712f7aca7a652254ba55#diff-c8606c3df3a0abbbcec36afcda70e4ce

tijoytom commented:
@gkhanna79 thanks for updating corlib.xml , i want to rename the (https://github.com/dotnet/coreclr/pull/5016) ZeroFreeAllocMemUTF8 to ZeroFreeCoTaskMemUTF8 , is it too late now? I can update model.corlib.xml but it will conflict with you PR.

gkhanna79 commented:
@tijoytom Please wait until I merge this PR (at noon today). Post that, please do the needful for both the model.xml files.

tijoytom commented:
@yizhang82  @gkhanna79 ok thanks . Will wait for you to merge.

mkborg commented:
After this change tests started to fail on Linux ARM since 'coreclr/tests/runtest.sh' script still copies 'mscorlib.dll' into 'coreoverlay' instead of 'System.Private.CoreLib.dll':
> open(".../coreclr/bin/tests/Windows_NT.x64.Debug/Tests/coreoverlay/System.Private.CoreLib.dll", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
